pennylane/workflow/qnode.py

```diff
@@ -837,8 +840,12 @@ def construct(self, args, kwargs):  # pylint: disable=too-many-branches
         params = self.tape.get_parameters(trainable_only=False)
         self.tape.trainable_params = qml.math.get_trainable_indices(params)
 
-        if isinstance(self._qfunc_output, qml.numpy.ndarray):
-            measurement_processes = tuple(self.tape.measurements)
+        if isinstance(self._qfunc_output, qml.numpy.tensor):
+            measurement_processes = tuple(
+                self.tape.measurements,
+            )
+        elif isinstance(self._qfunc_output, dict):
+            measurement_processes = tuple(self._qfunc_output.values())
         elif not isinstance(self._qfunc_output, Sequence):
             measurement_processes = (self._qfunc_output,)

```

Comment:
Perhaps here we can switch this to `measurement_processes = qml.pytrees.flatten(self._qfunc_output)[0]`.

Reply:
Potentially we could switch to storing `self._pytree_structure` instead of `self._qfunc_output`.  Basically we just need to store information somehow to remember how to postprocess the results.

Reply:
Yes, that could be an improvement. However, the typing of `qfunc_output` would also need to be stored as well since it is used to type convert results. 😓